### ISSUE TITLE
Add glob expressions to include directive to include multiple files

### ISSUE THREAD

**Issue by lordofthejars** [asciidoctor/asciidoctor#855]
I think it should be useful to add glob expression on include directive. It can be implemented as extension but then it will override the default behavior for all inclusions.

**Comment by mojavelinux** [comment]
Or perhaps you mean including multiple files?

**Comment by wolandscat** [comment]
Well the topic of this issue is a 'solution' not a problem statement. The problem i want to solve is including multiple files, for example as generated by a UML tool extractor, or any other code generator. If the code generate creates files in some structure like:

```
packageA
    +---packageA1
              +---classA1-1.adoc
              +---classA1-2.adoc
              +---...
              +---classA1-26.adoc
    +---packageA2
              +---classA2-1.adoc
packageB
    +---classB-1.adoc
    +---classB-2.adoc
```

The I want to be able to write something like 
```
== Package A
some text...

include::{uml_export_dir}/classes/packageA/*.adoc[leveloffset=+1]

=== Package A1
some other text
include::{uml_export_dir}/classes/packageA/packageA1/*.adoc[leveloffset=+1]

etc
```

or any syntax that has the intended meaning. At the moment, I have to do manual checking to make sure I have included all the files, which change due to changes in the model.

**Comment by mojavelinux** [comment]
Now I finally get the purpose of this issue. It's to use a single include directive to include multiple files.

The problem is, in what order? That's another dimension that needs to be expressed, further complicating the issue. That's why I think this really falls into the realm of an IncludeProcessor extension. That's exactly why we made that an extension point. Perhaps we can explore this idea in the extensions lab, find out how useful it is, then consider it for core once we iron out the design.

**Comment by AlexanderZobkov** [comment]
The feature can also be used for preparing a release notes document, where each new feature, fixed or know issue is presened by a separate asciidoc file (for modularity purpose). And you don't want to add a new include directive to the main document every time when a file describing a new feature or known issues is added by developers or tech writers. 
Also Asciidoc files could be automaticially generated from a bug tracker.

**Comment by wolandscat** [comment]
Order is the issue. I can see that one day it would be nice for certain documents that alphabetic or date ordering would be good. Where other ordering does matter (e.g. semantic). This isn't a must-have for us right now.

**Comment by mojavelinux** [comment]
Exactly, the open issue here is ordering of the matches. I think the key is just to define it. Alphabetically based on file system rules is probably the best default. What's most important is that we declare it to be predictable.

The other issue is going to be spacing between files. We can't rely on space between includes, so that brings #318 back into play.

**Comment by wimdeblauwe** [comment]
I would love to be able to have this. It would be very useful when working with ADRs (https://github.com/joelparkerhenderson/architecture_decision_record).

Each ADR is a separate adoc file. I would like to have a "master" adoc file that would create a book of each adoc file without having to edit the master each time a new file is added in a directory.

**Comment by mojavelinux** [comment]
It's very unlikely we'll implement this in core because there are too many different ways in which it could work, and trying to support all those ways would just make the include directive unnecessarily complex.

This is better done as an include processor extension. And, in fact, there's a very primitive version of such an extension available in the lab. See https://github.com/asciidoctor/asciidoctor-extensions-lab/blob/master/lib/glob-include-processor/extension.rb
